ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Background.hs

```diff
@@ -535,7 +535,11 @@ addBlockRunner
 addBlockRunner cdb@CDB{..} = forever $ do
     let trace = traceWith cdbTracer . TraceAddBlockEvent
     trace $ PoppedBlockFromQueue RisingEdge
-    blkToAdd <- getBlockToAdd cdbBlocksToAdd
-    trace $ PoppedBlockFromQueue $ FallingEdgeWith $
-      blockRealPoint $ blockToAdd blkToAdd
-    addBlockSync cdb blkToAdd
+    -- if the `addBlockSync` does not complete because it was killed by an async
+    -- exception (or it errored), notify the blocked thread)
+    bracketOnError (getBlockToAdd cdbBlocksToAdd)
+                   (\blkToAdd -> atomically $ tryPutTMVar (varBlockProcessed blkToAdd) Nothing)
+                   (\blkToAdd -> do
+                     trace $ PoppedBlockFromQueue $ FallingEdgeWith $
+                             blockRealPoint $ blockToAdd blkToAdd
+                     addBlockSync cdb blkToAdd)
```

Comment:
`bracketOnError` masks exceptions when `getBlockToAdd` blocks, this is ok (since `getBlockToAdd` blocks and thus is interruptible); once `getBlockToAdd` returns if there's an async exception that kills the thread it will guarantee that the `varBlockProcessed` `TMVar` is filled with `Nothing` (unless it was already filled by `addBlockSync`).

Reply:
Hmm. Do we also need the shutdown logic of the `addBlockRunner` thread to flush the `cdbBlocksToAdd` queue, sending `Nothing` to every request it hadn't started processing before it begin terminating?

Reply:
> do we ever restart the ChainDB? The queue is created when we start the ChainDB, if it's has the same life time as ChainDB then it doesn't matter, new ChainDB will anyway use new queue. But I suspect we never re-open ChainDB, is this right?

I'm pretty sure we never restart the `ChainDB`. (Maybe in tests?)

- Suppose the queue has two elements, `[A, B]`. (Recall that the block forging thread and the BlockFetch clients add to this queue concurrently.)
- Suppose `addBlockRunner` pops `B`, but the system goes down before it's done.
- If `A` was added by the block forging thread, then the code currently in the diff would never send `Nothing` to `A`'s promise; and so the block forging thread would be a zombie.

Reply:
Right, we indeed need to notify all the threads which posted a message in the queue.

Reply:
Fixed.